### PR TITLE
Hmu37: Added change for services with a PULL_REQUEST release cycle

### PR DESCRIPTION
**Pull request by jdvega-tuenti-com**
- Added support to make Jira seach using a jql
- Mapped linked issues when fetching a ticket
- Fetch Integration Request tickets from Release ticket 
- Use Gson to deserialize jira request

### CODE REVIEW

File: src/main/java/com/tuenti/acalvo/releasespy/service/TicketProvider.kt [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                .filter { versionAsNumber(it.version!!) >= previousVersion }
+                .filterNot { it.linkedTickets == null }
+                .flatMap { it.linkedTickets!! }
+                .map  { jiraHandler.getIssue(it.key) }
+                .toList()
+                .applyFilterConfig(ConfigHandler.instance.getConfig(upgrade).first(), upgrade)
+    }
+
+    private fun versionAsNumber(semanticVersion: String): Int =
+            semanticVersion.split(".").map { it.toInt() }.sum()
```
Comment (jdvega-tuenti-com) [inline src/main/java/com/tuenti/acalvo/releasespy/service/TicketProvider.kt]: I think I have a better way of calculating this...

```javascript
function versionAsNumber(v) {
    const vI = v.split(".").map(v => +v)
    let sum = 0
    for (var i = 0; i < vI.length; i++) {
        sum += vI[i] * Math.pow(10, vI.length - i - 1)
    }
    return sum
}
```

Any better idea? ( I just write it down on browser that's why I have used js)

File: src/main/java/com/tuenti/acalvo/releasespy/MsgHandler.kt [hunk 2]
```diff
@@ -128,26 +124,15 @@ class MsgHandler(private val jiraCredentials: JiraCredentials) {
...
-            .mapNotNull {
-                val res = TICKET.matchEntire(it)
-                if (res?.groupValues?.size == 2)
-                    res.destructured.component1()
-                else
-                    null
-            }
-            .distinct()
-            .map { JiraHandler(jiraCredentials).get(it) }
-            .filterNot { it.issueType == "Automated Release" }.toList()
```
Comment (jdvega-tuenti-com) [inline src/main/java/com/tuenti/acalvo/releasespy/MsgHandler.kt]: This code has been moved to TicketProvider

File: src/main/java/com/tuenti/acalvo/releasespy/MsgHandler.kt [hunk 3]
```diff
@@ -63,15 +59,15 @@ class MsgHandler(private val jiraCredentials: JiraCredentials) {
         edit(
             channel,
             params,
-            "Oops, I didn't get it. Example:`@releases-spy add my-service, brand-country/prod`") { e, s ->
+            "Oops, I didn't getIssue it. Example:`@releases-spy add my-service, brand-country/prod`") { e, s ->
```
Comment (jdvega-tuenti-com) [inline src/main/java/com/tuenti/acalvo/releasespy/MsgHandler.kt]:
```suggestion
            "Oops, I didn't get it. Example:`@releases-spy add my-service, brand-country/prod`") { e, s ->
```

File: src/main/java/com/tuenti/acalvo/releasespy/MsgHandler.kt [hunk 4]
```diff
@@ -63,15 +59,15 @@ class MsgHandler(private val jiraCredentials: JiraCredentials) {
...
             ConfigHandler.instance.add(channel, e, s)
         }
 
     private fun onRemove(channel: String, params: String) =
         edit(
             channel,
             params,
-            "Oops, I didn't get it. Example:`@releases-spy remove my-service, brand-country/prod`") { e, s ->
+            "Oops, I didn't getIssue it. Example:`@releases-spy remove my-service, brand-country/prod`") { e, s ->
```
Comment (jdvega-tuenti-com) [inline src/main/java/com/tuenti/acalvo/releasespy/MsgHandler.kt]:
```suggestion
            "Oops, I didn't get it. Example:`@releases-spy remove my-service, brand-country/prod`") { e, s ->
```

File: src/main/java/com/tuenti/acalvo/releasespy/ReleaseNotesScheduler.kt [hunk 5]
```diff
@@ -28,7 +28,7 @@ class ReleaseNotesScheduler {
                 "2.0",
                 LocalDateTime.of(2019, Month.JULY, 8, 11, 0),
                 listOf(
-                    ":female-cook: Bot self-configuration from Slack, use `@releases-spy help` to get more info",
+                    ":female-cook: Bot self-configuration from Slack, use `@releases-spy help` to getIssue more info",
```
Comment (polypiel) [inline src/main/java/com/tuenti/acalvo/releasespy/ReleaseNotesScheduler.kt]: Too much magic refactor

Comment (jdvega-tuenti-com) [reply src/main/java/com/tuenti/acalvo/releasespy/ReleaseNotesScheduler.kt]:
```suggestion
                    ":female-cook: Bot self-configuration from Slack, use `@releases-spy help` to get more info",
```

File: src/main/java/com/tuenti/acalvo/releasespy/JiraHandler.kt [hunk 6]
```diff
@@ -31,33 +62,49 @@ data class JiraTicket(
     val key: String,
     val summary: String? = null,
     val issueType: String? = null,
-    val assignee: String? = null)
+    val assignee: String? = null,
+    val reporter: String? = null,
```
Comment (polypiel) [inline src/main/java/com/tuenti/acalvo/releasespy/JiraHandler.kt]: Unused

Comment (jdvega-tuenti-com) [reply src/main/java/com/tuenti/acalvo/releasespy/JiraHandler.kt]:
```suggestion
```

File: src/main/java/com/tuenti/acalvo/releasespy/JiraHandler.kt [hunk 7]
```diff
@@ -1,25 +1,56 @@
...
+import java.util.*
+
 
 private data class Ticket(val key: String, val fields: Fields) {
-    fun ticket() = JiraTicket(key, fields.summary, fields.issuetype?.name, fields.assignee?.name)
+    fun ticket() = JiraTicket(key,
+            fields.summary,
+            fields.issuetype?.name,
+            fields.assignee?.name,
+            fields.reporter?.name,
```
Comment (jdvega-tuenti-com) [inline src/main/java/com/tuenti/acalvo/releasespy/JiraHandler.kt]:
```suggestion
```

File: src/main/java/com/tuenti/acalvo/releasespy/JiraHandler.kt [hunk 8]
```diff
@@ -1,25 +1,56 @@
...
+        it.ticket()
+    }
 }
-private data class Fields(val summary: String?, val issuetype: IssueType?, val assignee: Assignee? = null)
+
+private data class Fields(
+        val summary: String?,
+        val issuetype: IssueType?,
+        val assignee: User? = null,
+        val reporter: User? = null,
```
Comment (jdvega-tuenti-com) [inline src/main/java/com/tuenti/acalvo/releasespy/JiraHandler.kt]:
```suggestion
```

**Comment by jdvega-tuenti-com** [comment]
Ya he metido una función de `versionAsNumber` con cabeza :D

**Comment by jdvega-tuenti-com** [comment]
🎉

**Pull request merged**